text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,158 @@
+}
+    
+interface NotificationManager {
+  subscribeAttributeChanges(identifier, NotificationCallback)
+  subscribeRelationshipChanges(identifier, NotificationCallback)
+  subscribeErrorsChanges(identifier, NotificationCallback)
+  subscribeMetaChanges(identifier, NotificationCallback)
+  subscribeUnload(identifier, NotificationCallback)
+  unsubscribe(identifier, NotificationCallback)
+ }
+    
+class Store {
+  instantiateRecord(
+    identifier: RecordIdentifier, 
+    createArgs: CreateArgs, 
+    recordDataFor: RecordDataFor, 
+    notificationManager: NotificationManager): unknown)
+}
+```
+
+Instead of exposing the entire RecordData Class, we expose only the local facing methods and hide the server/adapter facing methods that the Model should not have access to, by giving a RecordData lookup method to `instantiateRecord`
+
+## Exposing schema information
+
+We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define these methods publicly on the store for addons to use. **These methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. The schema info is currently primarily geared towards being used by the internal relationship handling layer and the serializer/adapter layers.
+
+```ts
+export interface RelationshipSchema {
+  kind: 'hasMany'| 'belongsTo';
+  type: string;
+  options: { [key: string]: any } ;
+  name: string;
+}
+    
+interface RelationshipsDefinition {
+  [key: string]: RelationshipDefinition
+}
+    
+interface AttributeSchema {
+  name: string;
+  options: { [key: string]: any };
+  type: string;
+}
+
+interface AttributesDefinition {
+  [key: string]: AttributeSchema
+}
+    
+class Store {
+  // Following the existing RD implementation 
+  attributesDefitionFor(identifier: RecordIdentifier | type: string): AttributesDefiniton
+  
+  // Following the existing RD implementation
+  relationshipDefinitionFor(identifier: RecordIdentifier | type: string): RelationshipsDefinition
+}
+```
+    
+
+## Replacing Store methods that deal with DS.Model
+
+```ts
+class Store {
+  // deprecate
+  scheduleSave(model: DS.Model):Promise
+  reload(model: DS.Model): Promise
+  relationshipReferenceFor(model: DS.Model): Reference
+  
+  
+  // replace with 
+  scheduleSave(identifier: RecordIdentifier): Promise
+  reload(identifier: RecordIdentifier): Promise
+  relationshipReferenceFor(identifier: RecordIdentifier): Reference
+}
+```
+
+## Deprecating DS.Model being passed to serializers and adapters and store.modelFor
+
+Currently some adapter/serializer methods get the underlying class passed in. For example:
+
+```ts
+// Adapter
+createRecord(store, type, snapshot)
```

Comment:
I think it would be helpful to provide before & after examples here to show how to implement the changes required